File: code/notebooks/patchwork/action.py
from collections import namedtuple
from copy import deepcopy
from dataclasses import dataclass
from typing import Literal, Optional, Self

from .patch import Patch

Position = namedtuple('Position', 'row column')

@dataclass
class Action:
    """
    Represents an action that can be taken in the patchwork board game.
    """

    # ================================ static attributes ================================

    AMOUNT_OF_ACTIONS = 2026
    """
    The amount of available actions for the game of patchwork. The actually allowed actions are way lower than this number, but we need to be able to represent all the possible actions in a single number. This is the maximum amount of actions that can be taken in a single turn.

    (MAX_PATCH_INDEX(2) * ROWS(9) + MAX_ROW(8)) * COLUMNS(9) + MAX_COLUMN(8)) * ROTATIONS(4) + MAX_ROTATION(3)) * ORIENTATIONS(2) + MAX_ORIENTATION(1) + ACTIONS_OTHER_THAN_NORMAL_PATCH_PLACEMENT_ACTION(83)
    """

    # ================================ instance attributes ================================

    id: int
    """The id of the action. This is a number between 0 and 2025 (both inclusive)."""

    patch: Optional[Patch]
    """The patch that is being placed."""

    patch_position: Optional[Position]
    """The position of the patch that is being placed."""

    patch_index: Optional[Literal[0, 1, 2]]
    """The index of the patch from the list of all available patches."""

    # ================================ instance properties ================================

    @property
    def is_walking(self) -> bool:
        """Whether this action is a walking action."""
        return self.id == 0

    @property
    def is_special_patch_placement(self) -> bool:
        """Whether this action is a special patch placement action."""
        return 1 <= self.id <= 81

    @property
    def is_patch_placement(self) -> bool:
        """Whether this action is a normal patch placement action."""
        return 82 <= self.id <= 2025

    @property
    def is_first_patch_taken(self) -> bool:
        """Whether this action took the first available patch."""
        return 82 <= self.id < 730

    @property
    def is_second_patch_taken(self) -> bool:
        """Whether this action took the second available patch."""
        return 730 <= self.id < 1378

    @property
    def is_third_patch_taken(self) -> bool:
        """Whether this action took the third available patch."""
        return 1378 <= self.id < 2026

    # ================================ static methods ================================

    @staticmethod
    def walking() -> Self:
        """Returns a walking action."""
        return Action(None, None, None)

    # ================================ instance methods ================================

    def __init__(self, patch: Optional[Patch], patch_position: Optional[Position], patch_index: Optional[Literal[0, 1, 2]]):
        self.patch = patch
        self.patch_position = patch_position
        self.patch_index = patch_index

        ROWS = 9
        COLUMNS = 9
        ROTATIONS = 4
        ORIENTATIONS = 2

        if patch is None:
            # walking action [0, 0]
            self.id = 0
        elif patch is not None and patch_position is not None and patch_index is None:
            OFFSET = 1
            # special patch placement action [1, 81]
            self.id = patch_position.row * COLUMNS + patch_position.column + OFFSET
        else:
            OFFSET = 82
            # the maximum amount of placement for a patch is actually 448. The patch is:
            # ▉
            # ▉▉▉
            # but as we want to be able to represent all the information in a single number, we need to use [(((index * ROWS + row) * COLUMNS + column) * ROTATIONS + rotation) * ORIENTATIONS + orientation + OFFSET] as limit for the action
            self.id = (
                patch_index * ROWS * COLUMNS * ROTATIONS * ORIENTATIONS +
                patch_position.row * COLUMNS * ROTATIONS * ORIENTATIONS +
                patch_position.column        * ROTATIONS * ORIENTATIONS +
                patch.rotation.value                     * ORIENTATIONS +
                patch.orientation.value +
                OFFSET
            )

    def __repr__(self) -> str:
        return f'Action(id={self.id}, patch={self.patch}, patch_position={self.patch_position}, patch_index={self.patch_index})'

    def __str__(self) -> str:
        action_str = f'Action {self.id}'

        if self.is_walking:
            action_str += ' - Walking'
            return action_str

        if self.is_special_patch_placement:
            action_str += f' - Special patch placement at ({self.patch_position.row}, {self.patch_position.column})'
            return action_str

        action_str += f' - Patch placement of patch at index {self.patch_index} at ({self.patch_position.row}, {self.patch_position.column})'

        return action_str

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Action) and self.id == other.id

    def __hash__(self) -> int:
        return hash(self.id)

    def __copy__(self) -> Self:
        return Action(self.patch, self.patch_position, self.patch_index)

    def __deepcopy__(self, memo: dict) -> Self:
        return Action(deepcopy(self.patch, memo), deepcopy(self.patch_position, memo), self.patch_index)

    def copy(self) -> Self:
        return deepcopy(self)

File: code/patchwork/patchwork_core/src/action/action.rs
use std::{fmt::Display, hash::Hash};

use derivative::Derivative;

use crate::{Patch, QuiltBoard};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ActionPayload {
    /// The player is doing nothing. (null move)
    Null,
    /// The player is walking.
    Walking { starting_index: usize },
    /// The player is placing a patch.
    PatchPlacement {
        patch: &'static Patch,
        patch_index: usize,
        patch_rotation: usize,
        patch_orientation: usize,
        row: usize,
        column: usize,
        starting_index: usize,
        next_quilt_board: Option<u128>,
        previous_quilt_board: Option<u128>,
    },
    /// The player is placing a special patch.
    SpecialPatchPlacement {
        patch_id: usize,
        row: usize,
        column: usize,
        next_quilt_board: Option<u128>,
        previous_quilt_board: Option<u128>,
    },
}

/// Represents an action that can be taken in the patchwork board game.
#[derive(Debug, Clone, Derivative)]
#[derivative(PartialEq, Eq, Hash)]
pub struct Action {
    //  The id of the action. This is a number between 0 and 2025 (both inclusive).
    pub id: usize,
    /// The patch that is being placed or Walking if the player is walking.
    #[derivative(PartialEq = "ignore")]
    #[derivative(Hash = "ignore")]
    pub payload: ActionPayload,
}

impl Action {
    /// The amount of available actions for the game of patchwork.
    /// The actually allowed actions are way lower than this number,
    /// but we need to be able to represent all the possible actions in a single number.
    /// This is the maximum amount of actions that can be taken in a single turn.
    /// The actually best id is 2025, null moves have id 2026
    ///
    /// (MAX_PATCH_INDEX(2) * ROWS(9) + MAX_ROW(8)) * COLUMNS(9) + MAX_COLUMN(8)) * ROTATIONS(4) + MAX_ROTATION(3)) * ORIENTATIONS(2) + MAX_ORIENTATION(1) + ACTIONS_OTHER_THAN_NORMAL_PATCH_PLACEMENT_ACTION(83)
    pub const AMOUNT_OF_ACTIONS: u32 = 2026;

    /// The id of the null action.
    pub const NULL_ACTION_ID: usize = 2026;
    /// The id of the walking action.
    pub const WALKING_ACTION_ID: usize = 0;
    /// The id of the first special patch placement action.
    pub const SPECIAL_PATCH_PLACEMENT_ID_START: usize = 1;
    /// The id of the last special patch placement action.
    pub const SPECIAL_PATCH_PLACEMENT_ID_END: usize = 81;
    /// The id of the first patch placement action.
    pub const PATCH_PLACEMENT_ID_START: usize = 82;
    /// The id of the last patch placement action.
    pub const PATCH_PLACEMENT_ID_END: usize = 2025;

    /// Creates a new null `Action`.
    #[inline]
    pub fn null() -> Action {
        Action {
            id: Self::NULL_ACTION_ID,
            payload: ActionPayload::Null,
        }
    }

    // Creates a new walking `Action`.
    #[inline]
    pub fn walking(starting_index: usize) -> Action {
        Action {
            id: Self::WALKING_ACTION_ID,
            payload: ActionPayload::Walking { starting_index },
        }
    }

    #[inline]
    pub fn new(payload: ActionPayload) -> Action {
        Action {
            id: Action::calculate_id(&payload),
            payload,
        }
    }

    /// Whether this action is a null move.
    #[inline]
    pub fn is_null(&self) -> bool {
        self.id == usize::MAX
    }

    /// Whether this action is a walking action.
    #[inline]
    pub fn is_walking(&self) -> bool {
        self.id == 0
    }

    /// Whether this action is a special patch placement action.
    #[inline]
    pub fn is_special_patch_placement(&self) -> bool {
        matches!(self.payload, ActionPayload::SpecialPatchPlacement { .. })
    }

    /// Whether this action is a normal patch placement action.
    #[inline]
    pub fn is_patch_placement(&self) -> bool {
        matches!(self.payload, ActionPayload::PatchPlacement { .. })
    }

    /// Whether this action took the first patch.
    #[inline]
    pub fn is_first_patch_taken(&self) -> bool {
        match &self.payload {
            ActionPayload::PatchPlacement { patch_index, .. } => *patch_index == 0,
            _ => false,
        }
    }

    /// Whether this action took the second patch.
    #[inline]
    pub fn is_second_patch_taken(&self) -> bool {
        match &self.payload {
            ActionPayload::PatchPlacement { patch_index, .. } => *patch_index == 1,
            _ => false,
        }
    }

    /// Whether this action took the third patch.
    #[inline]
    pub fn is_third_patch_taken(&self) -> bool {
        match &self.payload {
            ActionPayload::PatchPlacement { patch_index, .. } => *patch_index == 2,
            _ => false,
        }
    }

    pub fn get_from_special_patch_placement_id(action_id: usize) -> (usize, usize, usize) {
        const OFFSET: usize = 1;
        const COLUMNS: usize = QuiltBoard::COLUMNS;

        let row = (action_id - OFFSET) / COLUMNS;
        let column = (action_id - OFFSET) % COLUMNS;
        let patch_id = action_id - OFFSET;

        (patch_id, row, column)
    }

    pub fn get_from_patch_placement_id(action_id: usize) -> (usize, usize, usize, usize, usize) {
        const OFFSET: usize = 82;
        const ROTATIONS: usize = 4;
        const ORIENTATIONS: usize = 2;
        const ROWS: usize = QuiltBoard::ROWS;
        const COLUMNS: usize = QuiltBoard::COLUMNS;

        let patch_index = (action_id - OFFSET) / (ROWS * COLUMNS * ROTATIONS * ORIENTATIONS);
        let row = ((action_id - OFFSET) / (COLUMNS * ROTATIONS * ORIENTATIONS)) % ROWS;
        let column = ((action_id - OFFSET) / (ROTATIONS * ORIENTATIONS)) % COLUMNS;
        let patch_rotation = ((action_id - OFFSET) / ORIENTATIONS) % ROTATIONS;
        let patch_orientation = (action_id - OFFSET) % ORIENTATIONS;

        (patch_index, row, column, patch_rotation, patch_orientation)
    }

    #[rustfmt::skip]
    fn calculate_id(payload: &ActionPayload) -> usize {
        const ROWS: usize = QuiltBoard::ROWS;
        const COLUMNS: usize = QuiltBoard::COLUMNS;

        match payload {
            // null action [usize::MAX, usize::MAX]
            ActionPayload::Null => {
                usize::MAX
            }
            // walking action [0, 0]
            ActionPayload::Walking { .. } => {
                0
            }
            // special patch placement action [1, 81]
            ActionPayload::SpecialPatchPlacement { row, column, .. } => {
                const OFFSET: usize = 1;
                *row * COLUMNS + *column + OFFSET
            }
            //
            // the maximum amount of placement for a patch is actually 448. The patch is:
            // ▉
            // ▉▉▉
            // but as we want to be able to represent all the information in a single number, we need to use [(((index * ROWS + row) * COLUMNS + column) * ROTATIONS + rotation) * ORIENTATIONS + orientation + OFFSET] as limit for the action
            ActionPayload::PatchPlacement { patch_index, row, column, patch_rotation, patch_orientation, .. } => {
                const OFFSET: usize = 82;
                const ROTATIONS: usize = 4;
                const ORIENTATIONS: usize = 2;

                *patch_index * ROWS * COLUMNS * ROTATIONS * ORIENTATIONS +
                       *row         * COLUMNS * ROTATIONS * ORIENTATIONS +
                       *column                * ROTATIONS * ORIENTATIONS +
                       *patch_rotation                    * ORIENTATIONS +
                       *patch_orientation +
                       OFFSET
            }
        }
    }
}

impl Display for Action {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Action {}", self.id)?;

        match &self.payload {
            ActionPayload::Null => {
                write!(f, " - NULL")
            }
            ActionPayload::Walking { starting_index } => {
                write!(f, " - Walking from index {}", starting_index)
            }
            ActionPayload::SpecialPatchPlacement {
                patch_id, row, column, ..
            } => {
                write!(
                    f,
                    " - Special patch({}) placement at ({}, {})",
                    *patch_id, *row, *column
                )
            }
            ActionPayload::PatchPlacement {
                patch,
                patch_index,
                row,
                column,
                patch_rotation,
                patch_orientation,
                ..
            } => {
                write!(
                    f,
                    " - Patch({}) placement (index {}) at ({}, {}) with (R {}, O {})",
                    patch.id, *patch_index, *row, *column, *patch_rotation, *patch_orientation
                )
            }
        }
    }
}
